File: implementation/scipamato/public/web/src/test/java/ch/difty/scipamato/publ/web/paper/browse/SimpleFilterPanelChangeEventTest.java
package ch.difty.scipamato.publ.web.paper.browse;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

import nl.jqno.equalsverifier.EqualsVerifier;
import nl.jqno.equalsverifier.Warning;
import org.apache.wicket.ajax.AjaxRequestTarget;
import org.apache.wicket.markup.html.form.TextArea;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class SimpleFilterPanelChangeEventTest {

    private SimpleFilterPanelChangeEvent e;

    @Mock
    private AjaxRequestTarget targetMock, targetMock2;

    @Mock
    private TextArea<?> mockComponent;

    @AfterEach
    void tearDown() {
        verifyNoMoreInteractions(targetMock, targetMock2);
    }

    @Test
    void canRetrieveTarget() {
        e = new SimpleFilterPanelChangeEvent(targetMock);
        assertThat(e.getTarget()).isEqualTo(targetMock);
    }

    @Test
    void usingMinimalConstructor_doesNotSetAnySpecialStuff() {
        e = new SimpleFilterPanelChangeEvent(targetMock);
        assertThat(e.getId()).isNull();
        assertThat(e.getMarkupId()).isNull();
    }

    @Test
    void usingWithId_doesAddId() {
        e = new SimpleFilterPanelChangeEvent(targetMock).withId("foo");
        assertThat(e.getId()).isEqualTo("foo");
        assertThat(e.getMarkupId()).isNull();
    }

    @Test
    void usingWithMarkupId_doesAddMarkupId() {
        e = new SimpleFilterPanelChangeEvent(targetMock).withMarkupId("bar");
        assertThat(e.getId()).isNull();
        assertThat(e.getMarkupId()).isEqualTo("bar");
    }

    @Test
    void usingWithIdAndMarkupId_doesAddBoth() {
        e = new SimpleFilterPanelChangeEvent(targetMock)
            .withId("hups")
            .withMarkupId("goo");
        assertThat(e.getId()).isEqualTo("hups");
        assertThat(e.getMarkupId()).isEqualTo("goo");
    }

    @Test
    void canOverrideTarget() {
        e = new SimpleFilterPanelChangeEvent(targetMock);
        assertThat(e.getTarget()).isEqualTo(targetMock);
        e.setTarget(targetMock2);
        assertThat(e.getTarget()).isEqualTo(targetMock2);
    }

    @Test
    void consideringAddingToTarget_withIdLessEvent_addsTarget() {
        e = new SimpleFilterPanelChangeEvent(targetMock);
        assertThat(e.getId()).isNull();

        e.considerAddingToTarget(mockComponent);

        verify(targetMock).add(mockComponent);
    }

    @Test
    void consideringAddingToTarget_withDifferingId_doesNotAddTarget() {
        e = new SimpleFilterPanelChangeEvent(targetMock)
            .withId("otherId")
            .withMarkupId("mId");
        e.considerAddingToTarget(mockComponent);
        verify(targetMock, never()).add(mockComponent);
    }

    @Test
    void consideringAddingToTarget_withSameIdButNullMarkupId_addsTarget() {
        when(mockComponent.getId()).thenReturn("id");
        when(mockComponent.getMarkupId()).thenReturn("mId");

        e = new SimpleFilterPanelChangeEvent(targetMock).withId("id");
        assertThat(e.getMarkupId()).isNull();

        e.considerAddingToTarget(mockComponent);

        verify(targetMock).add(mockComponent);
    }

    @Test
    void consideringAddingToTarget_withSameIdAndDifferingMarkupId_addsTarget() {
        when(mockComponent.getId()).thenReturn("id");
        when(mockComponent.getMarkupId()).thenReturn("mId");

        e = new SimpleFilterPanelChangeEvent(targetMock)
            .withId("id")
            .withMarkupId("otherMarkupId");
        e.considerAddingToTarget(mockComponent);
        verify(targetMock).add(mockComponent);
    }

    @Test
    void consideringAddingToTarget_withSameIdButSameMarkupId_doesNotAddTarget() {
        e = new SimpleFilterPanelChangeEvent(targetMock)
            .withId("id")
            .withMarkupId("mId");
        e.considerAddingToTarget(mockComponent);
        verify(targetMock, never()).add(mockComponent);
    }

    @Test
    void equals() {
        EqualsVerifier
            .forClass(SimpleFilterPanelChangeEvent.class)
            .withRedefinedSuperclass()
            .suppress(Warning.STRICT_INHERITANCE, Warning.NONFINAL_FIELDS)
            .verify();
    }

}

File: public/public-web/src/test/kotlin/ch/difty/scipamato/publ/web/paper/browse/SimpleFilterPanelChangeEventTest.kt
package ch.difty.scipamato.publ.web.paper.browse

import io.mockk.confirmVerified
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.verify
import nl.jqno.equalsverifier.EqualsVerifier
import nl.jqno.equalsverifier.Warning
import org.amshove.kluent.shouldBeEqualTo
import org.amshove.kluent.shouldBeNull
import org.apache.wicket.ajax.AjaxRequestTarget
import org.apache.wicket.markup.html.form.TextArea
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith

private const val ID = "id"
private const val MARKUP_ID = "mId"

@ExtendWith(MockKExtension::class)
internal class SimpleFilterPanelChangeEventTest {

    private lateinit var e: SimpleFilterPanelChangeEvent

    @MockK(relaxUnitFun = true)
    private lateinit var targetMock: AjaxRequestTarget

    @MockK(relaxUnitFun = true)
    private lateinit var targetMock2: AjaxRequestTarget

    @MockK
    private lateinit var mockComponent: TextArea<String>

    @BeforeEach
    fun setUp() {
        every { mockComponent.id } returns ID
        every { mockComponent.markupId } returns MARKUP_ID
    }

    @AfterEach
    fun tearDown() {
        confirmVerified(targetMock, targetMock2)
    }

    @Test
    fun canRetrieveTarget() {
        e = SimpleFilterPanelChangeEvent(targetMock)
        e.target shouldBeEqualTo targetMock
        verify { targetMock == targetMock }
    }

    @Test
    fun usingMinimalConstructor_doesNotSetAnySpecialStuff() {
        e = SimpleFilterPanelChangeEvent(targetMock)
        e.id.shouldBeNull()
        e.markupId.shouldBeNull()
    }

    @Test
    fun usingWithId_doesAddId() {
        e = SimpleFilterPanelChangeEvent(targetMock).withId("foo")
        e.id shouldBeEqualTo "foo"
        e.markupId.shouldBeNull()
    }

    @Test
    fun usingWithMarkupId_doesAddMarkupId() {
        e = SimpleFilterPanelChangeEvent(targetMock).withMarkupId("bar")
        e.id.shouldBeNull()
        e.markupId shouldBeEqualTo "bar"
    }

    @Test
    fun usingWithIdAndMarkupId_doesAddBoth() {
        e = SimpleFilterPanelChangeEvent(targetMock)
            .withId("hups")
            .withMarkupId("goo")
        e.id shouldBeEqualTo "hups"
        e.markupId shouldBeEqualTo "goo"
    }

    @Test
    fun canOverrideTarget() {
        e = SimpleFilterPanelChangeEvent(targetMock)
        e.target shouldBeEqualTo targetMock
        e.target = targetMock2
        e.target shouldBeEqualTo targetMock2
        verify { targetMock == targetMock }
        verify { targetMock2 == targetMock2 }
    }

    @Test
    fun consideringAddingToTarget_withIdLessEvent_addsTarget() {
        e = SimpleFilterPanelChangeEvent(targetMock)
        e.id.shouldBeNull()
        e.considerAddingToTarget(mockComponent)
        verify { targetMock.add(mockComponent) }
    }

    @Test
    fun consideringAddingToTarget_withDifferingId_doesNotAddTarget() {
        e = SimpleFilterPanelChangeEvent(targetMock)
            .withId("otherId")
            .withMarkupId(MARKUP_ID)
        e.considerAddingToTarget(mockComponent)
        verify(exactly = 0) { targetMock.add(mockComponent) }
    }

    @Test
    fun consideringAddingToTarget_withSameIdButNullMarkupId_addsTarget() {
        every { mockComponent.id } returns ID
        every { mockComponent.markupId } returns MARKUP_ID
        e = SimpleFilterPanelChangeEvent(targetMock).withId(ID)
        e.markupId.shouldBeNull()
        e.considerAddingToTarget(mockComponent)
        verify { targetMock.add(mockComponent) }
    }

    @Test
    fun consideringAddingToTarget_withSameIdAndDifferingMarkupId_addsTarget() {
        every { mockComponent.id } returns ID
        every { mockComponent.markupId } returns MARKUP_ID
        e = SimpleFilterPanelChangeEvent(targetMock)
            .withId(ID)
            .withMarkupId("otherMarkupId")
        e.considerAddingToTarget(mockComponent)
        verify { targetMock.add(mockComponent) }
    }

    @Test
    fun consideringAddingToTarget_withSameIdButSameMarkupId_doesNotAddTarget() {
        e = SimpleFilterPanelChangeEvent(targetMock)
            .withId(ID)
            .withMarkupId(MARKUP_ID)
        e.considerAddingToTarget(mockComponent)
        verify(exactly = 0) { targetMock.add(mockComponent) }
    }

    @Test
    fun equals() {
        EqualsVerifier
            .forClass(SimpleFilterPanelChangeEvent::class.java)
            .withRedefinedSuperclass()
            .suppress(Warning.STRICT_INHERITANCE, Warning.NONFINAL_FIELDS)
            .verify()
    }
}